jenkinsapi/plugins.py

```diff
@@ -57,6 +83,173 @@ def __getitem__(self, plugin_name):
+        """
+        Returns a list of all dependencies for a downloaded plugin
+        """
+        plugin_dependencies = []
+        manifest = read_manifest(downloaded_plugin)
+        manifest_dependencies = manifest.main_section.get('Plugin-Dependencies')
+        if manifest_dependencies:
+            dependencies = manifest_dependencies.split(',')
+            for dep in dependencies:
+                components = dep.split(';')  # split plugin:version;resolution:optional entries
+                dep_plugin = components[0]
+                name = dep_plugin.split(':')[0]
+                # install latest dependency, avoids multiple versions of the same dep
+                plugin_dependencies.append(Plugin({'shortName': name, 'version': 'latest'}))
+        return plugin_dependencies
+
+    def _download_plugin(self, download_link):
+        downloaded_plugin = StringIO()
+        downloaded_plugin.write(requests.get(download_link).content)
+        return downloaded_plugin
+
+    def _plugin_has_finished_installation(self, plugin):
+        """
+        Return True if installation is marked as 'Success' or 'SuccessButRequiresRestart'
+        in Jenkins' update_center, else return False.
+        """
+        try:
+            jobs = self.update_center_install_status['data']['jobs']
+            for job in jobs:
+                if job['name'] == plugin.shortName and job['installStatus'] in ['Success', 'SuccessButRequiresRestart']:
+                    return True
+            return False
+        except JenkinsAPIException:
+            return False  # lack of update_center in Jenkins 1.X
+
+    def plugin_version_is_being_installed(self, plugin):
+        """
+        Return true if plugin is currently being installed.
+        """
+        try:
+            jobs = self.update_center_install_status['data']['jobs']
+        except JenkinsAPIException:
+            return False  # lack of update_center in Jenkins 1.X
+        return any([job for job in jobs if job['name'] == plugin.shortName and job['version'] == plugin.version])
+
+    def plugin_version_already_installed(self, plugin):
+        """
+        Check if plugin version is already installed
+        """
+        if plugin.shortName not in self:
+            if self.plugin_version_is_being_installed(plugin):
+                return True
+            return False
+        installed_plugin = self[plugin.shortName]
+        if plugin.version == installed_plugin.version:
+            return True
+        elif plugin.version == "latest":
+            # we don't have an exact version, we first check if Jenkins knows about an update
+            if hasattr(installed_plugin, 'hasUpdates') and installed_plugin.hasUpdates:
+                return False
+            else:
+                # Jenkins may not have an up-to-date catalogue, so check update-center directly
+                latest_version = self.update_center_dict['plugins'][plugin.shortName]['version']
```

Comment:
I think that if Jenkins doesn't have up-to-date catalogue and there is no way to ask Jenkins to update itself - it is a bug in Jenkins and needs to be reported to CloudBees. It could be that call to refresh catalogue is not documented anywhere, but does exist.